connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -0,0 +1,416 @@
+    throw new UnsupportedOperationException
+  }
+
+  /**
+   * (Java-specific) Applies the given function to each sorted cogrouped data. For each unique
+   * group, the function will be passed the grouping key and 2 sorted iterators containing all
+   * elements in the group from [[Dataset]] `this` and `other`. The function can return an
+   * iterator containing elements of an arbitrary type which will be returned as a new
+   * [[Dataset]].
+   *
+   * This is equivalent to [[KeyValueGroupedDataset#cogroup]], except for the iterators to be
+   * sorted according to the given sort expressions. That sorting does not add computational
+   * complexity.
+   *
+   * @see
+   *   [[org.apache.spark.sql.KeyValueGroupedDataset#cogroup]]
+   * @since 3.5.0
+   */
+  def cogroupSorted[U, R](
+      other: KeyValueGroupedDataset[K, U],
+      thisSortExprs: Array[Column],
+      otherSortExprs: Array[Column],
+      f: CoGroupFunction[K, V, U, R],
+      encoder: Encoder[R]): Dataset[R] = {
+    cogroupSorted(other)(thisSortExprs: _*)(otherSortExprs: _*)(
+      UdfUtils.coGroupFunctionToScalaFunc(f))(encoder)
+  }
+}
+
+/**
+ * This class is the implementation of class [[KeyValueGroupedDataset]]. This class memorizes the
+ * initial types of the grouping function so that the original function will be sent to the server
+ * to perform the grouping first. Then any type modifications on the keys and the values will be
+ * applied sequentially to ensure the final type of the result remains the same as how
+ * [[KeyValueGroupedDataset]] behaves on the server.
+ */
+private class KeyValueGroupedDatasetImpl[K, V, IK, IV](
+    private val ds: Dataset[IV],
+    private val sparkSession: SparkSession,
+    private val plan: proto.Plan,
+    private val ikEncoder: AgnosticEncoder[IK],
+    private val kEncoder: AgnosticEncoder[K],
+    private val groupingFunc: IV => IK,
+    private val valueMapFunc: IV => V)
+    extends KeyValueGroupedDataset[K, V] {
+
+  private val ivEncoder = ds.encoder
+
+  override def keyAs[L: Encoder]: KeyValueGroupedDataset[L, V] = {
+    new KeyValueGroupedDatasetImpl[L, V, IK, IV](
+      ds,
+      sparkSession,
+      plan,
+      ikEncoder,
+      encoderFor[L],
+      groupingFunc,
+      valueMapFunc)
+  }
+
+  override def mapValues[W: Encoder](valueFunc: V => W): KeyValueGroupedDataset[K, W] = {
+    new KeyValueGroupedDatasetImpl[K, W, IK, IV](
+      ds,
+      sparkSession,
+      plan,
+      ikEncoder,
+      kEncoder,
+      groupingFunc,
+      valueMapFunc.andThen(valueFunc))
+  }
+
+  override def keys: Dataset[K] = {
+    ds.map(groupingFunc)(ikEncoder)
+      .dropDuplicates()
+      .as(kEncoder)
+  }
+
+  override def flatMapSortedGroups[U: Encoder](sortExprs: Column*)(
+      f: (K, Iterator[V]) => TraversableOnce[U]): Dataset[U] = {
+    // Apply mapValues changes to the udf
+    val nf: (K, Iterator[IV]) => TraversableOnce[U] =
+      UdfUtils.mapValuesAdaptor(f, valueMapFunc)
```

Comment:
A bit of OCD. We only we would need this step when there is an actual valueMapFunc right? We could skip when it is identity right?